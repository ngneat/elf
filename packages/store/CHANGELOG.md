# Changelog

This file was generated using [@jscutlery/semver](https://github.com/jscutlery/semver).

# [2.1.0](https://github.com/ngneat/elf/compare/store-2.0.0...store-2.1.0) (2022-05-30)


### Features

* **store:** add  setPropsInitialValue to have ability update initial value passed to propsFactory ([#275](https://github.com/ngneat/elf/issues/275)) ([c8be676](https://github.com/ngneat/elf/commit/c8be676dbf8418bb075630ef049f5c16c86bb1f5))



# [2.0.0](https://github.com/ngneat/elf/compare/store-1.5.6...store-2.0.0) (2022-05-12)


### Bug Fixes

* **store:** destroy should reset the store state ([558e4c0](https://github.com/ngneat/elf/commit/558e4c0e24b63ddfbe35162725d6c7bcc24d45a0))


### BREAKING CHANGES

* **store:** Calling the destroy method will also reset the state



## [1.5.6](https://github.com/ngneat/elf/compare/store-1.5.5...store-1.5.6) (2022-03-29)


### Bug Fixes

* **store:** üêû return emitOnce cb value ([4718f37](https://github.com/ngneat/elf/commit/4718f3766b0c89851b506301a9c6c1c4be52ddb9))



## [1.5.5](https://github.com/ngneat/elf/compare/store-1.5.4...store-1.5.5) (2022-03-28)


### Bug Fixes

* **store:** üêû fix emitOnce condition ([46d2bf4](https://github.com/ngneat/elf/commit/46d2bf4a7f284ae117783330ab44857a918fad1e))



## [1.5.4](https://github.com/ngneat/elf/compare/store-1.5.3...store-1.5.4) (2022-03-23)


### Bug Fixes

* **store:** add proper export for enableElfProdMode ([#203](https://github.com/ngneat/elf/issues/203)) ([190aab5](https://github.com/ngneat/elf/commit/190aab5df8cdabe4f39e6d7481fa596e35715a62))

## [1.5.3](https://github.com/ngneat/elf/compare/store-1.5.2...store-1.5.3) (2022-03-22)


### Bug Fixes

* **store:** add enableElfProdMode



## [1.5.2](https://github.com/ngneat/elf/compare/store-1.5.0...store-1.5.2) (2022-03-16)


### Bug Fixes

* **store:** fixes incorrect operator imports from rxjs ([#197](https://github.com/ngneat/elf/issues/197)) ([c8764f8](https://github.com/ngneat/elf/commit/c8764f8ae04bbe5aa89520f329461232d90f7d88))



# [1.5.0](https://github.com/ngneat/elf/compare/store-1.4.0...store-1.5.0) (2022-03-08)


### Features

* **store:** üî• add emitOnce function ([3256d5f](https://github.com/ngneat/elf/commit/3256d5fbd1603ded8365c01c518862b3b6d6f3bf))


# [1.4.0](https://github.com/ngneat/elf/compare/store-1.3.3...store-1.4.0) (2022-03-08)


### Features

* **store:** üî• elfHooks with preStoreUpdate hook ([#182](https://github.com/ngneat/elf/issues/182)) ([0704c33](https://github.com/ngneat/elf/commit/0704c3399c58008fa33702276943832a54d5dd49))



## [1.3.3](https://github.com/ngneat/elf/compare/store-1.3.2...store-1.3.3) (2022-02-24)


### Bug Fixes

* **store:** üêû fix createStore type ([f1f38b2](https://github.com/ngneat/elf/commit/f1f38b26157ca237c1cb60f67ce055bab627fef7))



## [1.3.2](https://github.com/ngneat/elf/compare/store-1.3.1...store-1.3.2) (2022-02-23)


### Bug Fixes

* **store:** üêû revert type ([370bb9c](https://github.com/ngneat/elf/commit/370bb9cfb33b98dd88058469297e070ddaf2f83d))



## [1.3.1](https://github.com/ngneat/elf/compare/store-1.3.0...store-1.3.1) (2022-02-23)


### Bug Fixes

* **store:** üêû fix createStore type ([87c756b](https://github.com/ngneat/elf/commit/87c756b79c47cc0d225f3bc4633ae50a3363ef9e))



# [1.3.0](https://github.com/ngneat/elf/compare/store-1.2.0...store-1.3.0) (2022-02-23)


### Features

* **store:** üî• introduce createStore function ([#174](https://github.com/ngneat/elf/issues/174)) ([fbc68ff](https://github.com/ngneat/elf/commit/fbc68ff1dd91190486a74dea9637ce34a47fb9ea))



# [1.2.0](https://github.com/ngneat/elf/compare/store-1.1.0...store-1.2.0) (2022-02-17)


### Features

* **store:** üî• add setProps mutation ([452d2aa](https://github.com/ngneat/elf/commit/452d2aa56df7097aff4b387eee97702b0b1c4f24))



# [1.1.0](https://github.com/ngneat/elf/compare/store-1.0.2...store-1.1.0) (2022-02-16)


### Features

* **store:** üî• add setProp mutation ([b0bea61](https://github.com/ngneat/elf/commit/b0bea61e67e476559a58ce35018460f8bfac68fb))



## [1.0.2](https://github.com/ngneat/elf/compare/store-1.0.1...store-1.0.2) (2021-12-14)



## [1.0.1](https://github.com/ngneat/elf/compare/store-1.0.0...store-1.0.1) (2021-12-14)


### Bug Fixes

* **entities:** typed support for immutable/readonly arrays ([#86](https://github.com/ngneat/elf/issues/86)) ([9cd6638](https://github.com/ngneat/elf/commit/9cd66381b7b9562eda10c52cd63bc19017ec8bbb))
