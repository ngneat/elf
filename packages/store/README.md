# @ngneat/elf

[Docs](https://ngneat.github.io/elf/)

