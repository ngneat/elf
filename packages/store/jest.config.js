module.exports = {
  displayName: 'store',
  preset: '../../jest.preset.js',
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    },
  },
  testEnvironment: 'node',
  transform: {
    '^.+\\.[tj]sx?$': 'ts-jest',
  },
  testPathIgnorePatterns: ['stale'],
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/packages/store',
};
