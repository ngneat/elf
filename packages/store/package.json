{
  "name": "@ngneat/elf",
  "version": "1.0.2",
  "description": "A reactive state management for JS applications",
  "publishConfig": {
    "access": "public"
  },
  "bugs": {
    "url": "https://github.com/ngneat/elf/issues"
  },
  "homepage": "https://github.com/ngneat/elf#readme",
  "repository": {
    "type": "git",
    "url": "https://github.com/ngneat/elf"
  },
  "keywords": [
    "rxjs",
    "state management",
    "reactive state management",
    "js state management",
    "reactive store",
    "elf"
  ],
  "author": {
    "name": "Netanel Basal",
    "email": "netanel7799@gmail.com",
    "url": "https://netbasal.com"
  },
  "license": "MIT",
  "sideEffects": false,
  "peerDependencies": {
    "typescript": ">= 4.1.2"
  }
}
