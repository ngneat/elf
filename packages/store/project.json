{
  "root": "packages/store",
  "sourceRoot": "packages/store/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nrwl/web:rollup",
      "outputs": ["{options.outputPath}"],
      "options": {
        "updateBuildableProjectDepsInPackageJson": false,
        "outputPath": "dist/packages/store",
        "tsConfig": "packages/store/tsconfig.lib.json",
        "project": "packages/store/package.json",
        "entryFile": "packages/store/src/index.ts",
        "globals": [
          {
            "global": "Rx",
            "moduleId": "rxjs"
          },
          {
            "global": "Rx",
            "moduleId": "rxjs/operators"
          }
        ],
        "rollupConfig": "@nrwl/react/plugins/bundle-rollup",
        "assets": [
          {
            "glob": "README.md",
            "input": ".",
            "output": "."
          }
        ]
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": ["packages/store/**/*.ts"]
      },
      "outputs": ["{options.outputFile}"]
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/packages/store"],
      "options": {
        "jestConfig": "packages/store/jest.config.js",
        "passWithNoTests": true
      }
    },
    "version": {
      "executor": "@jscutlery/semver:version",
      "options": {
        "syncVersions": false
      }
    }
  },
  "tags": []
}
