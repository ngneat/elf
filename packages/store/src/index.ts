export {
  Store,
  Reducer,
  StoreValue,
  StoreDef,
  ReducerContext,
} from './lib/store';
export { propsFactory } from './lib/props-factory';
export { propsArrayFactory } from './lib/props-array-factory';
export {
  select,
  head,
  asap,
  filterNil,
  distinctUntilArrayItemChanged,
} from './lib/operators';
export {
  getStore,
  getRegistry,
  registry$,
  getStoresSnapshot,
} from './lib/registry';
export { withProps } from './lib/props.state';
export { createState, PropsFactory, EmptyConfig } from './lib/state';
export {
  isFunction,
  coerceArray,
  isObject,
  isString,
  isUndefined,
  capitalize,
} from './lib/utils';
export { StateOf, Query, OrArray } from './lib/types';
