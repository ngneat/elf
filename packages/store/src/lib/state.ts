type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never;
type Merge<State extends any[], Key extends PropertyKey> = UnionToIntersection<
  State[number][Key]
>;

export type PropsFactory<Props, Config> = { props: Props; config: Config };
export type EmptyConfig = undefined;

export function createState<S extends PropsFactory<any, any>[]>(
  ...propsFactories: S
): { state: Merge<S, 'props'>; config: Merge<S, 'config'> } {
  const result = { config: {}, state: {} };

  for (const { config, props } of propsFactories) {
    Object.assign(result.config, config);
    Object.assign(result.state, props);
  }

  return result as { state: Merge<S, 'props'>; config: Merge<S, 'config'> };
}
