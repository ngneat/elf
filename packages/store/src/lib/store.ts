import { BehaviorSubject, Observable } from 'rxjs';
import { Query } from '..';
import { addStore, removeStore } from './registry';

export class Store<
  SDef extends StoreDef = any,
  State = SDef['state']
> extends BehaviorSubject<State> {
  initialState!: State;

  private context: ReducerContext = {
    config: this.getConfig(),
  };

  constructor(private storeDef: SDef) {
    super(storeDef.state);
    this.initialState = this.getValue();
    addStore(this);
  }

  get name(): StoreDef['name'] {
    return this.storeDef.name;
  }

  getConfig<Config extends Record<any, any>>(): Config {
    return this.storeDef.config;
  }

  query<R>(selector: Query<State, R>) {
    return selector(this.getValue());
  }

  update(...reducers: Array<Reducer<State>>) {
    const currentState = this.getValue();

    const nextState = reducers.reduce((value, reducer) => {
      value = reducer(value, this.context);

      return value;
    }, currentState);

    if (nextState !== currentState) {
      super.next(nextState);
    }
  }

  reset() {
    this.update(() => this.initialState);
  }

  combine<O extends Record<string, Observable<any>>>(
    observables: O
  ): Observable<{
    [P in keyof O]: O[P] extends Observable<infer R> ? R : never;
  }> {
    let hasChange = true;
    const buffer: Record<string, any> = {};

    return new Observable((observer) => {
      for (const [key, query] of Object.entries(observables)) {
        observer.add(
          query.subscribe((value) => {
            buffer[key] = value;
            hasChange = true;
          })
        );
      }

      return this.subscribe({
        next() {
          if (hasChange) {
            observer.next(buffer as any);
            hasChange = false;
          }
        },
        error(e) {
          observer.error(e);
        },
        complete() {
          observer.complete();
        },
      });
    });
  }

  destroy() {
    removeStore(this);
  }

  next(value: State) {
    this.update(() => value);
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  error() {}

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  complete() {}
}

export type StoreValue<T extends Store> = ReturnType<T['getValue']>;
export type Reducer<State> = (state: State, context: ReducerContext) => State;
export type ReducerContext = { config: Record<PropertyKey, any> };

export interface StoreDef<State = any> {
  name: string;
  state: State;
  config: any;
}
