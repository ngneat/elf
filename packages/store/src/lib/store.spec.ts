import {
  addEntities,
  selectAllEntities,
  UIEntitiesRef,
  updateEntities,
  withEntities,
  withUIEntities,
} from '@ngneat/elf-entities';
import { createTodo, createUITodo, Todo } from '@ngneat/elf-mocks';
import { withProps } from './props.state';
import { createState } from './state';
import { Store } from './store';

type UIEntity = { id: number; open: boolean };

describe('store', () => {
  it('should reset store', () => {
    const { state, config } = createState(
      withEntities<Todo>(),
      withProps<{ filter: string }>({ filter: '' })
    );

    const store = new Store({ state, name: 'todos', config });

    store.update(addEntities(createTodo(1)), (state) => ({
      ...state,
      filter: 'foo',
    }));

    expect(store.getValue()).toMatchSnapshot();

    store.reset();

    expect(store.getValue()).toMatchSnapshot();
  });

  it('should combine and fire only once', () => {
    const { state, config } = createState(
      withEntities<Todo>(),
      withUIEntities<UIEntity>(),
      withProps<{ filter: string }>({ filter: '' })
    );

    const store = new Store({ state, name: 'todos', config });

    const spy = jest.fn();

    store
      .combine({
        todos: store.pipe(selectAllEntities()),
        ui: store.pipe(selectAllEntities({ ref: UIEntitiesRef })),
      })
      .subscribe(spy);

    expect(spy).toHaveBeenCalledTimes(1);

    expect(store.getValue()).toMatchSnapshot();

    store.update(
      addEntities(createTodo(1)),
      addEntities(createUITodo(1), { ref: UIEntitiesRef })
    );

    expect(spy).toHaveBeenCalledTimes(2);
    expect(store.getValue()).toMatchSnapshot();

    store.update((state) => ({ ...state, filter: 'foo' }));

    // Update non related value should not call `next`
    expect(spy).toHaveBeenCalledTimes(2);

    store.update(updateEntities(1, { title: 'foo' }), (state) => ({
      ...state,
      filter: 'hello',
    }));

    expect(store.getValue()).toMatchSnapshot();
    expect(spy).toHaveBeenCalledTimes(3);
  });
});
