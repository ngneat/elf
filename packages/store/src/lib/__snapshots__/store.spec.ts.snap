// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store should combine and fire only once 1`] = `
Object {
  "UIEntities": Object {},
  "UIIds": Array [],
  "entities": Object {},
  "filter": "",
  "ids": Array [],
}
`;

exports[`store should combine and fire only once 2`] = `
Object {
  "UIEntities": Object {
    "1": Object {
      "id": 1,
      "open": false,
    },
  },
  "UIIds": Array [
    1,
  ],
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
  },
  "filter": "",
  "ids": Array [
    1,
  ],
}
`;

exports[`store should combine and fire only once 3`] = `
Object {
  "UIEntities": Object {
    "1": Object {
      "id": 1,
      "open": false,
    },
  },
  "UIIds": Array [
    1,
  ],
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "foo",
    },
  },
  "filter": "hello",
  "ids": Array [
    1,
  ],
}
`;

exports[`store should reset store 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "completed": false,
      "id": 1,
      "title": "todo 1",
    },
  },
  "filter": "foo",
  "ids": Array [
    1,
  ],
}
`;

exports[`store should reset store 2`] = `
Object {
  "entities": Object {},
  "filter": "",
  "ids": Array [],
}
`;
